src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -81,6 +82,105 @@ def _impute_not_reviewed_labels(user, messages_traced_data, analysis_dataset_con
+
+
+def _impute_ws_coding_errors(user, messages_traced_data, analysis_dataset_configs, ws_correct_dataset_code_scheme):
+    """
+    Imputes Codes.CODING_ERROR labels for messages that have a coding error in the WS labels that have been applied.
+
+    We consider WS labels to have a coding error if either of these conditions holds:
+     - There is a WS label  applied in a normal code scheme, but there is no label in the WS - Correct Dataset code scheme.
+     - There is a label applied in the WS - Correct Dataset code scheme, but no WS code in any of the normal code schemes.
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to impute not reviewed labels.
+    :type messages_traced_data: list of TracedData
+    :param analysis_dataset_configs: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
+    :param ws_correct_dataset_code_scheme: WS - Correct Dataset code scheme.
+    :type ws_correct_dataset_code_scheme: core_data_modules.data_models.CodeScheme
+    """
+    log.info(f"Imputing {Codes.CODING_ERROR} labels for WS codes...")
+    imputed_labels = 0
+    for message_td in messages_traced_data:
+        message = Message.from_dict(dict(message_td))
+
+        message_analysis_config = analysis_dataset_config_for_message(analysis_dataset_configs, message)
+        normal_code_schemes = [c.code_scheme for c in message_analysis_config.coding_configs]
+
+        # Check for a WS code in any of the normal code schemes
+        ws_code_in_normal_scheme = False
+        for label in message.get_latest_labels():
+            if not label.checked:
+                continue
+
+            if label.scheme_id != ws_correct_dataset_code_scheme.scheme_id:
+                code = _code_for_label(label, normal_code_schemes)
+                if code.control_code == Codes.WRONG_SCHEME:
+                    ws_code_in_normal_scheme = True
+
+        # Check for a code in the WS code scheme
+        code_in_ws_scheme = False
+        for label in message.get_latest_labels():
+            if not label.checked:
+                continue
+
+            if label.scheme_id == ws_correct_dataset_code_scheme.scheme_id:
+                code_in_ws_scheme = True
+
+        if ws_code_in_normal_scheme != code_in_ws_scheme:
+            imputed_labels += 1
+
+            # Clear all duplicate schemes
+            valid_code_scheme_ids = [code_scheme.scheme_id for code_scheme in normal_code_schemes]
+            valid_code_scheme_ids.append(ws_correct_dataset_code_scheme.scheme_id)
+            for label in message.get_latest_labels():
+                cleared_label = None
+                for scheme_id in valid_code_scheme_ids:
+                    if label.scheme_id.startswith(scheme_id):
```

Comment:
Sorry, I might have missed this - but why are we inserting a cleared_label here?

Reply:
The messages at this point contain labels in Coda's format. If we want to say that these labels are invalid and a coding error, we need to replace the currently assigned labels across all schemes with Coding Error, taking care to clear pre-existing codes in duplicated schemes, in the same way that a human labeller would clear codes in Coda. Clearing all the labels here ensures that only the new labels we write will be returned by latest_labels() later on

Reply:
added an explanation as a comment in the code